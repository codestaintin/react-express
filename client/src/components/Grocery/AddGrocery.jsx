import React from 'react';
import PropTypes from 'prop-types';

/**
 * @class AddGrocery
 *
 * @returns {JSX}
 */
class AddGrocery extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      name: ''
    };
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleChange = this.handleChange.bind(this);
  }

  handleSubmit(event) {
    event.preventDefault();
    this.props.createGrocery(this.state.name);
    this.setState({ name: '' });
  }

  handleChange(event) {
    this.setState({ name: event.target.value });
  }

  render() {
    const { name } = this.state;
    return (
      <div className="container">
        <form
          className="form-inline custom_form"
          onSubmit={this.handleSubmit}
        >
          <div className="form-group mx-sm-3 mb-2">
            <input
              className="form-control"
              name="name"
              placeholder="Add Grocery"
              pattern="^[a-zA-z ]*$"
              type="text"
              required
              onChange={this.handleChange}
              value={name}
            />
          </div>
          <button
            type="submit"
            className="btn btn-outline-primary mb-2"
          >
            Add Grocery
          </button>
        </form>
      </div>
    );
  }
}

AddGrocery.propTypes = {
  createGrocery: PropTypes.func.isRequired
};

export default AddGrocery;
