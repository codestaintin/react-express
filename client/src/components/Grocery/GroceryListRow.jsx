import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

const GroceryListRow = ({
  grocery, handleDelete, handlePurchase
}) => (
  <Fragment>
    <tr>
      <th scope="row">&nbsp;</th>
      <td>{grocery.name}</td>
      <td>
        <button
          type="button"
          className={`btn ${grocery.purchased ? 'btn-info' : 'btn-primary'}`}
          onClick={() => handlePurchase(grocery._id)}
        >
          {grocery.purchased ? 'Bought' : 'Buy'}
        </button>
      </td>
      <td>
        <button
          type="button"
          className="btn btn-danger"
          onClick={() => handleDelete(grocery._id)}
        >
              Delete Item
        </button>
      </td>
    </tr>
  </Fragment>
);

GroceryListRow.propTypes = {
  grocery: PropTypes.shape({}),
  handleDelete: PropTypes.func.isRequired,
  handlePurchase: PropTypes.func.isRequired
};

export default GroceryListRow;
