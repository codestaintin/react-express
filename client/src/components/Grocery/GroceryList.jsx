import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import GroceryListRow from './GroceryListRow';

const GroceryList = ({ groceries, handleDelete, handlePurchase }) => (
  <Fragment>
    <div className="container">
      <table className="table table-hover main-section">
        <thead>
          <tr>
            <th scope="col">&nbsp;</th>
            <th scope="col">Grocery Name</th>
            <th scope="col">Purchase Status</th>
            <th scope="col">Action</th>
          </tr>
        </thead>
        <tbody>
          {
            groceries.map(grocery => (
              <GroceryListRow
                key={grocery._id}
                grocery={grocery}
                handleDelete={handleDelete}
                handlePurchase={handlePurchase}
              />
            ))
          }
        </tbody>
      </table>
    </div>
  </Fragment>
);

GroceryList.propTypes = {
  groceries: PropTypes.array.isRequired,
  handleDelete: PropTypes.func.isRequired,
  handlePurchase: PropTypes.func.isRequired
};

export default GroceryList;
