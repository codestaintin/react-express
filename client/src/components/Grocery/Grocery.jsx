import React, { Fragment } from 'react';
import PropTypes, { shape } from 'prop-types';
import { connect } from 'react-redux';
import Loader from 'react-loader-spinner';
import { bindActionCreators } from 'redux';
import {
  getAllGroceries,
  deleteGrocery,
  createGrocery,
  purchaseGrocery
} from '../../actions/grocery/groceryActions';
import AddGrocery from './AddGrocery';
import GroceryList from './GroceryList';

/**
 * @class Grocery
 *
 * @returns {JSX}
 */
class Grocery extends React.Component {
  constructor(props) {
    super(props);
    this.handleDelete = this.handleDelete.bind(this);
    this.handleCreateGrocery = this.handleCreateGrocery.bind(this);
    this.handlePurchase = this.handlePurchase.bind(this);
  }

  componentDidMount() {
    this.props.getAllGroceries();
  }

  handleDelete(id) {
    this.props.deleteGrocery(id);
  }

  handleCreateGrocery(grocery) {
    this.props.createGrocery(grocery);
  }

  handlePurchase(id) {
    this.props.purchaseGrocery(id);
  }

  render() {
    const { groceries, loading } = this.props;
    return (
      <Fragment>
        <AddGrocery createGrocery={this.handleCreateGrocery} />
        {loading
        && (
          <div className="d-flex justify-content-center">
            <Loader type="Audio" color="#007bff" className="text-center" />
          </div>
        )
        }
        {
          !loading && groceries.length > 0
            && (
              <GroceryList
                groceries={groceries}
                handleDelete={this.handleDelete}
                handlePurchase={this.handlePurchase}
              />
            )
        }

        {
          !loading && !groceries.length > 0
              && (
                <h1 className="text-center">
                Please add new items
                </h1>
              )
        }
      </Fragment>

    );
  }
}

const mapStateToProps = ({ allGroceries }) => ({
  groceries: allGroceries.groceries,
  loading: allGroceries.isLoading
});

const mapDispatchToProps = dispatch => bindActionCreators({
  getAllGroceries,
  deleteGrocery,
  createGrocery,
  purchaseGrocery
}, dispatch);

Grocery.propTypes = {
  groceries: PropTypes.arrayOf(shape({})),
  getAllGroceries: PropTypes.func.isRequired,
  deleteGrocery: PropTypes.func.isRequired,
  createGrocery: PropTypes.func.isRequired,
  purchaseGrocery: PropTypes.func.isRequired,
  loading: PropTypes.bool.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(Grocery);
