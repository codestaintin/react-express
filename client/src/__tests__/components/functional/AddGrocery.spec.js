import React from 'react';
import { shallow } from 'enzyme';
import { AddGrocery } from '../../../components/Grocery/AddGrocery';

describe('<AddGrocery>', () => {
  const props = {
    createGrocery: jest.fn(() => {})
  };
  const tree = shallow(<AddGrocery {...props} />);

  it('should render AddGrocery', () => {
    expect(tree).toMatchSnapshot();
  });

  it('should simulate handleSubmit', () => {
    const spy = jest.spyOn(tree.instance(), 'handleSubmit');
    tree.setState({
      name: 'Onion'
    });
    tree.instance().handleSubmit(
      {
        preventDefault: jest.fn()
      }
    );
    expect(spy).toHaveBeenCalled();
  });

  it('should simulate handleChange', () => {
    const spy = jest.spyOn(tree.instance(), 'handleChange');
    tree.instance().handleChange({
      target: {
        name: ''
      }
    });
    expect(spy).toHaveBeenCalled();
  });
});