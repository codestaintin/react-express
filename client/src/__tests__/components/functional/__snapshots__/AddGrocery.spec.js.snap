// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddGrocery> should render AddGrocery 1`] = `
<div
  className="container"
>
  <form
    className="form-inline custom_form"
    onSubmit={[Function]}
  >
    <div
      className="form-group mx-sm-3 mb-2"
    >
      <input
        className="form-control"
        name="name"
        onChange={[Function]}
        pattern="^[a-zA-z ]*$"
        placeholder="Add Grocery"
        required={true}
        type="text"
        value=""
      />
    </div>
    <button
      className="btn btn-outline-primary mb-2"
      type="submit"
    >
      Add Grocery
    </button>
  </form>
</div>
`;
